File: lib/spiral/threading.ts
import { defaultOf, IDisposable } from "../../deps/Fable/src/fable-library-ts/Util.js";
import { union_type, class_type, TypeInfo } from "../../deps/Fable/src/fable-library-ts/Reflection.js";
import { Union } from "../../deps/Fable/src/fable-library-ts/Types.js";
import { Option } from "../../deps/Fable/src/fable-library-ts/Option.js";

export class Disposable implements IDisposable {
    readonly f: (() => void);
    constructor(f: (() => void)) {
        this.f = f;
    }
    Dispose(): void {
        const _: Disposable = this;
        _.f();
    }
}

export function Disposable_$reflection(): TypeInfo {
    return class_type("Threading.Disposable", undefined, Disposable);
}

export function Disposable_$ctor_3A5B6456(f: (() => void)): Disposable {
    return new Disposable(f);
}

export type US0_$union = 
    | US0<0>
    | US0<1>

export type US0_$cases = {
    0: ["US0_0", [any]],
    1: ["US0_1", []]
}

export function US0_US0_0(f0_0: any) {
    return new US0<0>(0, [f0_0]);
}

export function US0_US0_1() {
    return new US0<1>(1, []);
}

export class US0<Tag extends keyof US0_$cases> extends Union<Tag, US0_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US0_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US0_0", "US0_1"];
    }
}

export function US0_$reflection(): TypeInfo {
    return union_type("Threading.US0", [], US0, () => [[["f0_0", class_type("System.Threading.CancellationToken")]], []]);
}

export function closure1(unitVar: void, v0_1: any): US0_$union {
    return US0_US0_0(v0_1);
}

export function method0(): ((arg0: any) => US0_$union) {
    return (v: any): US0_$union => closure1(undefined, v);
}

export function closure2(v0_1: any, unitVar: void): void {
    defaultOf();
}

export function method1(v0_1: any): (() => void) {
    return (): void => {
        closure2(v0_1, undefined);
    };
}

export function method2(v0_1: any): (() => void) {
    return (): void => {
        closure2(v0_1, undefined);
    };
}

export function closure0(unitVar: void, v0_1: Option<any>): [any, IDisposable] {
    const patternInput: [any, IDisposable] = defaultOf();
    const _run_target_args$0027_v1 = [patternInput[0], patternInput[1]] as [any, IDisposable];
    return [_run_target_args$0027_v1[0], _run_target_args$0027_v1[1]] as [any, IDisposable];
}

export const v0 = (v: Option<any>): [any, IDisposable] => closure0(undefined, v);

export function new_disposable_token(x: Option<any>): [any, IDisposable] {
    return v0(x);
}

File: lib/spiral/threading_.py
from __future__ import annotations
from collections.abc import Callable
from typing import Any
from fable_modules.fable_library.async_builder import CancellationToken
from fable_modules.fable_library.reflection import (TypeInfo, class_type, union_type)
from fable_modules.fable_library.types import (Array, Union)
from fable_modules.fable_library.util import IDisposable

def _expr102() -> TypeInfo:
    return class_type("Threading.Disposable", None, Disposable)


class Disposable(IDisposable):
    def __init__(self, f: Callable[[], None]) -> None:
        self.f: Callable[[], None] = f

    def Dispose(self, __unit: None=None) -> None:
        _: Disposable = self
        _.f(None)


Disposable_reflection = _expr102

def Disposable__ctor_3A5B6456(f: Callable[[], None]) -> Disposable:
    return Disposable(f)


def _expr103() -> TypeInfo:
    return union_type("Threading.US0", [], US0, lambda: [[("f0_0", class_type("System.Threading.CancellationToken"))], []])


class US0(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US0_0", "US0_1"]


US0_reflection = _expr103

def closure1(unit_var: None, v0_1: CancellationToken) -> US0:
    return US0(0, v0_1)


def method0(__unit: None=None) -> Callable[[CancellationToken], US0]:
    def _arrow104(v: CancellationToken) -> US0:
        return closure1(None, v)

    return _arrow104


def closure2(v0_1: Any, unit_var: None) -> None:
    pass


def method1(v0_1: Any) -> Callable[[], None]:
    def _arrow105(__unit: None=None, v0_1: Any=v0_1) -> None:
        closure2(v0_1, None)

    return _arrow105


def method2(v0_1: Any) -> Callable[[], None]:
    def _arrow106(__unit: None=None, v0_1: Any=v0_1) -> None:
        closure2(v0_1, None)

    return _arrow106


def closure0(unit_var: None, v0_1: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    pattern_input: tuple[CancellationToken, IDisposable] = None
    _run_target_args_0027_v1: tuple[CancellationToken, IDisposable] = (pattern_input[0], pattern_input[1])
    return (_run_target_args_0027_v1[0], _run_target_args_0027_v1[1])


def _arrow107(v: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    return closure0(None, v)


v0: Callable[[CancellationToken | None], tuple[CancellationToken, IDisposable]] = _arrow107

def new_disposable_token(x: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    return v0(x)


